### DISCUSSION TITLE
How to correctly set up eglot and clangd formatting on Windows?

### DISCUSSION THREAD

**Discussion by bradJM** [joaotavora/eglot#672] [Q&A]
Hi everyone,

I'm not completely sure if this is an eglot issue or a problem with my configuration (leaning towards the latter, though). I'm using eglot with clangd on Windows, and I have a `.clang-format` file with `UseCRLF` set to `true`. Emacs reports my existing files as having DOS line endings, and no control characters show up in my buffers. After I use `eglot-format-buffer`, all my lines end with a visible `^M` character and the newlines are duplicated (ie an extra empty line between every existing line) if I check the file in another editor. I'm admittedly not very familiar with properly setting up Emacs different coding variables, but does this sound familiar to anyone?

My `init.el` contains:

```
(prefer-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
(w32-set-system-coding-system 'utf-8)
(setq-default buffer-file-coding-system 'utf-8)
```

But honestly, that's just based on some Googling and trial and error. 😅 I've also tried using `utf-8-dos` explicitly with the same result. Thank you for any help!

**Comment by joaotavora** [comment]
Hello! It might be a bug in eglot, yes. To my understanding Emacs normally shows those characters of there's at least one non CRLF line. You have to setup a minimal reproducible example and then open an issue.

**Reply by bradJM** [reply]
Okay, thanks! I assumed I just had it configured wrong, but if it could be a bug then I'll get a repro together and make an issue. 🙂

**Reply by joaotavora** [reply]
It _could_ be a bug, but I'd have to look at the event transcript, at least.

**Reply by joaotavora** [reply]
If you don't have all the elements to make an issue yet, don't.  Paste an LSP event transcript here instead.

**Reply by bradJM** [reply]
I went ahead and made a [quick reproduction](https://github.com/bradJM/eglot-crlf-repro) of it just to make sure I could reproduce it with a stripped down config. I put the event transcript in a [gist](https://gist.github.com/bradJM/b0652809cf0d605ea5ed87d8e9b74d73#file-gistfile1-txt-L131-L172) because it's quite long. The issue might be a bit clearer in this screenshot since the hidden characters are visible:

![image](https://user-images.githubusercontent.com/1667870/115553791-31cf8b00-a284-11eb-9e93-3d3fbad6f1ce.png)

One thing I find interesting: what I thought was happening (the buffer was being rewritten with both CRLF and LF) is _half_ true:

Before formatting:

```
$ file test.cpp
test.cpp: C source, ASCII text, with CRLF line terminators
```

and after:

```
$ file test.cpp
test.cpp: C source, ASCII text, with CRLF, CR line terminators
```
